### PR TITLE
Added support for register-with-taints setting #93608

### LINKED ISSUE #93608: Support for --register-with-taints in kubemark hollow-kubelet args
**What would you like to be added**:
The kubemark hollow-kubelet should expose setting --register-with-taints through args for the test.

**Why is this needed**:
1. It would be beneficial to have the nodes come up with taints to avoid scheduling of regular pods on the nodes. The specific workloads can then be scheduled on the hollow-nodes by adding tolerations. This would help for environments where the actual nodes and the kubemark nodes run on the same cluster.
2. Sometimes the hollow nodes are scheduled on other hollow-nodes making them unusable.

### PR DESCRIPTION
**Pull request by kshitijcode**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #93608

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**: 
Yes

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cmd/kubemark/hollow-node.go [hunk 1]
```diff
@@ -88,6 +89,9 @@ func (c *hollowNodeConfig) addFlags(fs *pflag.FlagSet) {
 	fs.DurationVar(&c.ProxierMinSyncPeriod, "proxier-min-sync-period", 0, "Minimum period that proxy rules are refreshed in hollow-proxy.")
 	bindableNodeLabels := cliflag.ConfigurationMap(c.NodeLabels)
 	fs.Var(&bindableNodeLabels, "node-labels", "Additional node labels")
+	bindableRegisterWithTaints := cliflag.ConfigurationMap(c.RegisterWithTaints)
+	fs.Var(&bindableRegisterWithTaints, "register-with-taints", "Adding Taints")
+
```
Comment (mm4tt) [inline cmd/kubemark/hollow-node.go:94]: Remove blank line

File: cmd/kubemark/hollow-node.go [hunk 2]
```diff
@@ -64,6 +64,7 @@ type hollowNodeConfig struct {
 	ProxierSyncPeriod    time.Duration
 	ProxierMinSyncPeriod time.Duration
 	NodeLabels           map[string]string
+	RegisterWithTaints   map[string]string
```
Comment (mm4tt) [inline cmd/kubemark/hollow-node.go:67]: You also need to propagate this to the kubelet flags, exactly like the NodeLabels is done - https://github.com/kubernetes/kubernetes/blob/8d54a84507c622e1552373d5d125c84ee22d1465/pkg/kubemark/hollow_kubelet.go#L164

Comment (kshitijcode) [reply cmd/kubemark/hollow-node.go:67]: @mm4tt resolved :)

**Comment by neolit123** [comment]
/sig scalability
are release notes needed for kubemark?

**Comment by kshitijcode** [comment]
/help 
Sorry, I am unaware but what are the next steps on this PR workflow?
I tried searching for the relevant documentation but couldn't find .

File: cmd/kubemark/hollow-node.go [hunk 3]
```diff
@@ -88,6 +89,8 @@ func (c *hollowNodeConfig) addFlags(fs *pflag.FlagSet) {
 	fs.DurationVar(&c.ProxierMinSyncPeriod, "proxier-min-sync-period", 0, "Minimum period that proxy rules are refreshed in hollow-proxy.")
 	bindableNodeLabels := cliflag.ConfigurationMap(c.NodeLabels)
 	fs.Var(&bindableNodeLabels, "node-labels", "Additional node labels")
+	bindableRegisterWithTaints := cliflag.ConfigurationMap(c.RegisterWithTaints)
+	fs.Var(&bindableRegisterWithTaints, "register-with-taints", "Adding Taints")
```
Comment (mm4tt) [inline cmd/kubemark/hollow-node.go:93]: Please change the help message to something like: "Register the hollow node with the given list of taints"

**Comment by mm4tt** [comment]
We don't need release notes for kubemark, please set them to none

/ok-to-test

**Comment by kshitijcode** [comment]
@mm4tt : I am not able to figure out what failed the tests. It doesn't seem to be related to my change.
Can you please confirm?

**Comment by kshitijcode** [comment]
@mm4tt : I am not sure about the reason for failures. They don't seem to be related to my change. Can you please confirm?

**Comment by kshitijcode** [comment]
@mm4tt 
What should be the default value when exposing RegisterWithTaints in &KubeletFlags

```return &KubeletFlags{
		ContainerRuntimeOptions: *NewContainerRuntimeOptions(),
		CertDirectory:           "/var/lib/kubelet/pki",
		RootDirectory:           defaultRootDir,
		MasterServiceNamespace:  metav1.NamespaceDefault,
		MaxContainerCount:       -1,
		MaxPerPodContainerCount: 1,
		MinimumGCAge:            metav1.Duration{Duration: 0},
		NonMasqueradeCIDR:       "10.0.0.0/8",
		RegisterSchedulable:     true,
		RemoteRuntimeEndpoint:   remoteRuntimeEndpoint,
		NodeLabels:              make(map[string]string),
		RegisterNode:            true,
		SeccompProfileRoot:      filepath.Join(defaultRootDir, "seccomp"),
		// prior to the introduction of this flag, there was a hardcoded cap of 50 images
		EnableCAdvisorJSONEndpoints: false,
		RegisterWithTaints: 	 []core.Taint,
	}```

File: cmd/kubemark/hollow-node.go (hunk 2 above)
Comment (kshitijcode) [reply cmd/kubemark/hollow-node.go:67]: @mm4tt : I changed it to type []core.Taint
Is it as intended?

**Comment by gongguan** [comment]
RegisterWithTaints is a slice not a map, you should replace `bindableRegisterWithTaints := cliflag.ConfigurationMap(c.RegisterWithTaints)` like this
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/options/options.go#L338

**Comment by kshitijcode** [comment]
@mm4tt @gongguan 
I have made the necessary change. I am not sure why the tests are failing. Locally it seems to be working fine.

**Comment by gongguan** [comment]
pass RegisterWithTaints from config to options here
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubemark/hollow-node.go#L109

**Comment by kshitijcode** [comment]
> https://github.com/kubernetes/kubernetes/blob/master/cmd/kubemark/hollow-node.go#L109

@gongguan I added it there.

What about adding it here as well https://github.com/kubernetes/kubernetes/blob/382107e6c84374b229e6188207ef026621286aa2/cmd/kubelet/app/options/options.go#L191 ?

**Comment by kshitijcode** [comment]
> pass RegisterWithTaints from config to options here
> https://github.com/kubernetes/kubernetes/blob/master/cmd/kubemark/hollow-node.go#L109

Also, how do we interpret the failed tests, some of the tests just say the failed error without much information?

**Comment by gongguan** [comment]
you should format changed files and run hack/update-bazel.sh

**Comment by kshitijcode** [comment]
@gongguan I did it. 
`hack/update-bazel.sh` didn't spit out an error.

**Comment by mm4tt** [comment]
The build is failing with:
```
W0806 08:27:38.879] Use --sandbox_debug to see verbose messages from the sandbox
W0806 08:27:38.880] compilepkg: missing strict dependencies:
W0806 08:27:38.880] 	/bazel-scratch/.cache/bazel/_bazel_prow/48d5366022b4e3197674c8d6e2bee219/sandbox/linux-sandbox/6785/execroot/io_k8s_kubernetes/pkg/kubemark/hollow_kubelet.go: import of "k8s.io/kubernetes/pkg/apis/core"
```

Yo need to add the deps to the bazel's BUILD files

**Comment by kshitijcode** [comment]
@mm4tt Do you mean adding 
``` "//pkg/apis/core:go_default_library",
        "//pkg/apis/core/pods:go_default_library",
        "//pkg/apis/core/v1:go_default_library",
        "//pkg/apis/core/v1/helper:go_default_library",
        "//pkg/apis/core/v1/helper/qos:go_default_library",```
to https://github.com/kubernetes/kubernetes/blob/master/pkg/kubemark/BUILD

**Comment by kshitijcode** [comment]
@mm4tt : added the dependencies, not really surely if its at the right place.

**Comment by gongguan** [comment]
> @mm4tt : added the dependencies, not really surely if its at the right place.
> Still doesn't seem to be fixed .

Run `./hack/update-bazel.sh`, BUILD files will be corrected automatically.

**Comment by kshitijcode** [comment]
@mm4tt @gongguan I fixed most of the errors. Some tests seem to failing.  Not really sure what the exact root cause is.

**Comment by gongguan** [comment]
> @mm4tt @gongguan I fixed most of the errors. Some tests seem to failing. Not really sure what the exact root cause is.

I think these are flaky tests which not related to your pr, try retest to succeed them.
You should squash your commits into one at first because it's misleading in git logs.

**Comment by kshitijcode** [comment]
@mm4tt Can you please help me with a `lgtm` ?

**Comment by mm4tt** [comment]
Sorry, this fell off my radar.

/lgtm
/approve

**Comment by kshitijcode** [comment]
@mm4tt : What would be the next steps to get this merged?

**Comment by mm4tt** [comment]
It will automatically get merged once the 1.19 code thaw is lifted. 
It will land into 1.20 release. The change is not critical enough to be part of 1.19

**Comment by kshitijcode** [comment]
@mm4tt : I see the thaw is lifted. Can this be merged now for the subsequent release?

**Comment by wojtek-t** [comment]
It will be picked automatically. The code thaw is not fully lifted - we're lifting that in phases.

**Comment by wojtek-t** [comment]
@kshitijcode - please rebase the PR (we're ready to merge it)

**Comment by wojtek-t** [comment]
Something didn't work - there are now 250 commits in it..

**Comment by kshitijcode** [comment]
Checking ..

**Comment by kshitijcode** [comment]
@wojtek-t  I am not able to figure out just did a git-rebase. Can you help me with the same?

**Comment by wojtek-t** [comment]
I'm not able to fix your PR - you need to do it yourself.

**Comment by kshitijcode** [comment]
@wojtek-t Sorry, I think something went wrong with my git history- reforked and  raised a fresh PR.

**Pull request closed without merging**